param_passing/proposal.md

```diff
@@ -0,0 +1,194 @@
+
+## Example
+
+Let's look at an example, here is an original model `base_robot`:
+
+```xml
+<model name="base_robot">
+  ...
+  <link name="chassis">
+    <visual name="lidar_visual">
+      ...
+    </visual>
+    <sensor name="lidar">
+      ...
+    </sensor>
+    <sensor name="camera">
+      <pose>0.5 0.02 0 0 0</pose>
+      ...
+    </sensor>
+  </link>
+
+  <link name="top">
+    <visual name="camera_visual">
+      ...
+      <geometry>
+        <box>
+          <size>0.03 0.03 0.05</size>
+        </box>
+      </geometry>
+      <material>
+        <ambient>0.0 1.0 0.0 1</ambient>
+        <diffuse>0.0 1.0 0.0 1</diffuse>
+        <specular>0.5 0.5 0.5 1</specular>
+      </material>
+    </visual>
+    <sensor name="camera">
+      <pose>0.2 1 0.5 0 0 0</pose>
+      <update_rate>20</update_rate>
+      ...
+    </sensor>
+  </link>
+```
+
+Now the user would like to use the original model `base_robot` in a world file but with modifications. The modifications include removing the lidar from the chassis link, adding a plugin to the camera sensor in the chassis link, adding a camera visual to the chassis link, update camera information in the top link, and replace the top link's camera visual geometry.
+
+```xml
+<model name="custom_robot">
+  <include>
+    <uri>model://base_robot</uri>
+
+    <experimental:params>
+      <visual name="chassis::lidar_visual" action="remove"/>
+      <sensor name="chassis::lidar" action="remove"/>
+
+     <sensor name="chassis::camera" action="add">
+        <plugin name="camera_depth_sensor" filename="libcamera_depth_sensor.so"/>
+     </sensor>
+
+      <visual name="chassis::camera_visual" action="add">
+        <pose>0.5 0.02 0 0 0</pose>
+        <geometry>
+            <box>
+                <size>0.02 0.02 0.02</size>
+            </box>
+        </geometry>
+      </visual>
+
+      <sensor name="top::camera" action="modify">
+        <pose>0 1 0 0 0 0</pose>
+        <update_rate>60</update_rate>
+      </sensor>
+
+      <visual name="top::camera_visual">
+        <geometry action="replace">
+          <sphere>
+            <radius>0.05</radius>
+          </sphere>
+        </geometry>
+        <material action="modify">
+          <ambient>0.0 1.0 0.0 1</ambient>
+        </material>
+      </visual>
+
+    </experimental:params>
+  </include>
+</model>
+```
+
+In `//include/experimental::params`, the action can be either listed in the element identifier or listed individually in the children element(s). For example,
+
+```xml
+<sensor name="top::camera" action="modify">
+  <pose>0 1 0 0 0 0</pose>
+  <update_rate>60</update_rate>
+</sensor>
+```
+
+The action is defined at the top level where the `name` identifier is, which means that all children listed in this element will be modified. If the user wants to do several different actions to the children elements then the user can specify the action for each child element individually. For instance,
```

Comment:
Would the `action` attribute be expected only on the immediate children of the tag with the `name` identifier? For example, could we have 
```xml
<visual name="top::camera_visual">
  <geometry>
    <sphere action="replace">
       <radius>0.05</radius>
    </sphere>
  </geometry>
</visual>
```

Reply:
I think the easiest way is to say that the `action` attribute is expected either in the direct children of `//include` where the `element_id` is or in the direct children of the element identifier. If we allow it to be anywhere (like in your example) I believe this might be more error prone. What are your thoughts?

Reply:
I agree. I think limiting it the locations you mentioned keeps the parsing code simpler without losing too much functionality--a trade-off I'm willing to make.